py/desispec/io/raw.py

```diff
@@ -68,7 +68,7 @@ def read_raw(filename, camera, fibermapfile=None, **kwargs):
     try:
         tmp = int(header['NIGHT'])
     except (KeyError, ValueError, TypeError):
-        header['NIGHT'] = header2night(header)
+        header['NIGHT'] = primary_header['NIGHT']
```

Comment:
This need to go back to `header2night(header)` or maybe `header2night(primary_header)` or something using `header2night` to be able to catch all the special cases of missing HEADER, str vs. int, ...

Reply:
I would like to be able to preprocess `/global/cfs/cdirs/desi/spectro/teststand/rawdata/SM01/2018/20180403/WINLIGHT_00007873.fits`  camera `z1` , but hdu `Z1` does not have date or mjd info. It's only in the primary header.

Reply:
Cascading chains of try/except: I updated it so that if header2night(header) also fails, it will fall back to using the primary header.  This supports both 20210105 with corrupted NIGHT keywords, normal nights with good NIGHT keywords, and early teststand data which only have NIGHT / timestamp keywords in the primary header.  I tested your case with
```
desi_preproc -i /global/cfs/cdirs/desi/spectro/teststand/rawdata/SM01/2018/20180403/WINLIGHT_00007873.fits --cameras z1 --outfile $SCRATCH/blat.fits
```